kratos/utilities/assign_master_slave_constraints_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,242 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main author:     Sebastian Ares de Parga Regalado
+//
+
+// System includes
+#include <string>
+#include <iostream>
+
+// External includes
+#include "concurrentqueue/concurrentqueue.h"
+
+// Project includes
+#include "utilities/assign_master_slave_constraints_to_neighbours_utility.h"
+
+namespace Kratos 
+{
+///@addtogroup Kratos Core
+///@{
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @class AssignMasterSlaveConstraintsToNeighboursUtility
+ * @ingroup Kratos Core
+ * @brief Assing Master-Slave Constraints to Neighbouring Nodes
+ * @details This class provides a method to search for neighbouring nodes of one node
+ * within a given radius and assign a master-slave constraint using radial basis functions.
+ * @author Sebastian Ares de Parga Regalado
+ */
+
+/// Default constructor.
+AssignMasterSlaveConstraintsToNeighboursUtility::AssignMasterSlaveConstraintsToNeighboursUtility(NodesContainerType& rMasterStructureNodes){
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_model_part = rMasterStructureNodes.GetContainer();
+    mpBins = Kratos::make_unique<NodeBinsType>(nodes_model_part.begin(), nodes_model_part.end());
+    mMaxNumberOfNodes = rMasterStructureNodes.size();
+    KRATOS_CATCH(""); 
+} 
+
+// Destructor
+AssignMasterSlaveConstraintsToNeighboursUtility::~AssignMasterSlaveConstraintsToNeighboursUtility() {}
+
+// Search for neighboring nodes (in rMasterStructureNodes) of the given slave node
+void AssignMasterSlaveConstraintsToNeighboursUtility::SearchNodesInRadiusForNode(
+    NodeType::Pointer pSlaveNode,
+    double const Radius,
+    ResultNodesContainerType& rCloudOfNodes)
+{
+
+    KRATOS_TRY;
+
+    ResultNodesContainerType local_results(mMaxNumberOfNodes);
```

Comment:
it would be better if you pass the local_results in the argument, making it a TLS variable if it is in a parallel loop